hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java

```diff
@@ -1460,6 +1472,64 @@ SnapshotDiffReport decodeResponse(Map<?, ?> json) {
     }.run();
   }
 
+  public SnapshotDiffReport getSnapshotDiffReport(final Path snapshotDir,
+      final String fromSnapshot, final String toSnapshot) throws IOException {
+    statistics.incrementReadOps(1);
+    storageStatistics.incrementOpCounter(OpType.GET_SNAPSHOT_DIFF);
+
+    if (!isValidSnapshotName(fromSnapshot) || !isValidSnapshotName(toSnapshot)) {
+      // In case the diff needs to be computed between a snapshot and the current
+      // tree, we should not do iterative diffReport computation as the iterative
+      // approach might fail if in between the rpc calls the current tree
+      // changes in absence of the global fsn lock.
+      return getSnapshotDiffReportWithoutListing(snapshotDir, fromSnapshot, toSnapshot);
+    } else {
+      byte[] startPath = DFSUtilClient.EMPTY_BYTES;
+      int index = -1;
+      SnapshotDiffReportGenerator snapshotDiffReport;
+      List<DiffReportListingEntry> modifiedList = new TreeList();
+      List<DiffReportListingEntry> createdList = new ChunkedArrayList<>();
+      List<DiffReportListingEntry> deletedList = new ChunkedArrayList<>();
+      SnapshotDiffReportListing report;
+
+      do {
+        try {
+          report = new FsPathResponseRunner<SnapshotDiffReportListing>(
+              GetOpParam.Op.GETSNAPSHOTDIFFLISTING,
+              snapshotDir,
+              new OldSnapshotNameParam(fromSnapshot),
+              new SnapshotNameParam(toSnapshot),
+              new SnapshotDiffStartPathParam(DFSUtilClient.bytes2String(startPath)),
+              new SnapshotDiffIndexParam(index)) {
+            @Override
+            SnapshotDiffReportListing decodeResponse(Map<?, ?> json) {
+              return JsonUtilClient.toSnapshotDiffReportListing(json);
+            }
+          }.run();
+        } catch (UnsupportedOperationException e) {
+          // In case the server doesn't support getSnapshotDiffReportListing,
```

Comment:
it would be nice to cache and remember this server does not support getSnapshotDiffReportListing.
Maybe as a follow-up enhancement.

Reply:
Let me do this in follow-up JIRA. I filed [HDFS-16249](https://issues.apache.org/jira/browse/HDFS-16249).